### PR TITLE
fix: Prevent lock race and content loss on language switch

### PR DESCRIPTION
**Pull request by afbora**
## Description

Builds on #8310 — same goal, just moved to a safer spot so it doesn't create new problems.

Before, the lock was released inside `unlock()`, which waited for the save first. That had two side effects: on tab close / refresh the unlock request sometimes never went out, and on a quick language switch a late save could put the lock back again.

Now `unlock()` stays simple, and any pending changes are saved on the language switch itself — before the lock is released. So nothing gets lost, and the lock stays released. If that save fails, we just show the error and don't switch, instead of leaving things half-done.

Also added tests for the language switch.

## Testing

Tested all three cases both with an E2E run and manually, all passing: the lock is released when switching language, the lock no longer comes back after a slow save, and a failed save cleanly aborts the switch instead of losing anything.

### 🐛 Bug fixes
- Release the content lock reliably when switching the translation language, without dropping unsaved changes #8310

### 🧹 Housekeeping
- Add tests for the language switch button (lock release + edge cases)

### CODE REVIEW

File: panel/src/components/View/Buttons/LanguagesDropdown.vue [hunk 1]
```diff
@@ -80,6 +80,18 @@ export default {
...
+			// switching the language cannot drop them or leave the lock behind
+			// due to a save that finishes after the unlock request
+			if (this.$panel.content.hasDiff() === true) {
+				try {
+					await this.$panel.content.update();
+				} catch (error) {
+					this.$panel.error(error);
+					return;
+				}
+			}
```
Comment (distantnative) [inline panel/src/components/View/Buttons/LanguagesDropdown.vue]: Shouldn't this rather live in the unlock method? Cause I would think that should always happen when unlocking, not only when changing languages, no?

Comment (afbora) [reply panel/src/components/View/Buttons/LanguagesDropdown.vue]: @distantnative Tried that first, but `await`-ing the save inside `unlock()` breaks tab close/refresh — the page dies before the `sendBeacon` fires, so the lock never releases. That's the exact bug this fixes, so the save has to stay on the language-switch side where we can actually wait for it.

Happy to change it if you see a cleaner way though.

Comment (distantnative) [reply panel/src/components/View/Buttons/LanguagesDropdown.vue]: @bastianallgeier @afbora I still think this should not be solved individually at each caller side but inside `content.unlock()`. Maybe we need a flag then for that, calling `content.unlock()` with something like `instant`/`nonblocking` or so and set that flag for the one specific caller that relies on no await to use `sendBeacon`.

Comment (afbora) [reply panel/src/components/View/Buttons/LanguagesDropdown.vue]: @distantnative I think the tricky part is that `unlock()` isn't only called from the language switch. `destroyed` and `beforeunload` in `ModelView` also call it without awaiting (through a small wrapper), so they rely on the beacon and can't really wait for a save.

So if the save moves into `unlock()`, I guess it would have to both save and tell the caller whether it worked, but that "did it save" answer only really makes sense for the one caller that can actually wait. Probably why it ended up on the language-switch side.

Not fully sure though, can you double check too please?

Comment (bastianallgeier) [reply panel/src/components/View/Buttons/LanguagesDropdown.vue]: I've added a new commit, which breaks up the unlock methods into two new methods: unlockBeaconRequest and unlockPostRequest. This helps quite nicely to pull the unsaved changes code into the unlock method, while still being able to call unblockBeaconRequest directly from the ModelView.unlock method.

Comment (afbora) [reply panel/src/components/View/Buttons/LanguagesDropdown.vue]: This looks a lot more elegant, nice one 👍 I'll leave the code approval to @distantnative though 🙂 

One question: should the error messages go through translation?

Comment (bastianallgeier) [reply panel/src/components/View/Buttons/LanguagesDropdown.vue]: Yes, I think it should be translated, but wanted to wait first if you like it :)

File: panel/src/panel/content.js [hunk 2]
```diff
@@ -275,6 +275,13 @@ export default (panel) => {
 
 		/**
 		 * Saves any changes
+		 *
+		 * Locked and aborted requests are handled here instead of being
+		 * thrown, as most callers save in the background and cannot react
+		 * to them. The return value tells callers that need to know
+		 * whether the changes actually made it to the server.
```
Comment (distantnative) [inline panel/src/panel/content.js]: I think this is phrased too much in relation what was there before or what could have been. Instead of stating the current state clearly

```
Locked and aborted requests return as false.
```

Maybe not even needed considering we also add a `@return` description.

File: panel/src/panel/content.js [hunk 3]
```diff
@@ -339,38 +339,70 @@ export default (panel) => {
...
+				// request took over. Staying on the current view keeps
+				// both the changes and the lock, so nothing is lost and
+				// the call can simply be repeated
+				if ((await this.update({}, env)) !== true) {
+					throw new Error("The changes could not be saved before unlocking");
+				}
+			}
+
+			// fail silently because the lock will be released after
+			// 10 minutes anyway
```
Comment (distantnative) [inline panel/src/panel/content.js]: I wouldn't state 10 minutes here - that's controlled somewhere else in the code and if we change that, this comment will be wrong and overlooked easily.

File: panel/src/panel/content.js [hunk 4]
```diff
@@ -339,38 +339,70 @@ export default (panel) => {
...
+			return this.unlockPostRequest(env).catch(() => {});
+		},
 
+		/**
+		 * Sends the unlock request for the given view.
+		 * Use sendBeacon for reliability during page unload. Browsers
+		 * guarantee delivery even when the page is being closed.
+		 */
+		unlockBeaconRequest(env = {}) {
```
Comment (distantnative) [inline panel/src/panel/content.js]: I would suggest a name that is more about the intent, less the technical implementation. Maybe `tryUnlock` or `unlockForgiving`?

File: panel/src/panel/content.test.ts [hunk 5]
```diff
@@ -0,0 +1,222 @@
+import { afterEach, beforeEach, describe, expect, it, vi } from "vitest";
+import Content from "./content.js";
+
```
Comment (distantnative) [inline panel/src/panel/content.test.ts:26]: It would be great if we could create this directly as `.ts` file, so that we can hopefully merge it without the biggest headaches when merging into `v6/develop`. Right now, it'll easily be overlooked that on v6/develop it would stick around as `.js` next to the `.ts` file.

It would also allow us to have proper parameter types right off the start instead of the slightly odd multiple docblock @param lines.

File: panel/src/panel/content.js (hunk 4 above)
Comment (bastianallgeier) [reply panel/src/panel/content.js]: I wanted to keep the naming consistent between unlockBeaconRequest and unlockPostRequest. I'm also not super convinced by the alternatives tbh.

Comment (distantnative) [reply panel/src/panel/content.js]: ok

File: panel/src/panel/content.test.ts (hunk 5 above)
Comment (bastianallgeier) [reply panel/src/panel/content.test.ts:26]: I gave it a try to let it migrate to ts, but I'm not sure if this is actually what we want. @distantnative you are the one who's deepest into typescript. Is this leading anywhere?

Comment (bastianallgeier) [reply panel/src/panel/content.test.ts:26]: @distantnative could you have another look at this?

Comment (distantnative) [reply panel/src/panel/content.test.ts:26]: It's a little tricky as now also `content.ts` got migrated to TypeScript, but not 100% like we have it on `v6/develop`. Trying to figure out the best way to deal with merge conflicts then and not to regress parts from this fix and/or from the `v6/develop` state when merging into `v6/develop`.

Comment (distantnative) [reply panel/src/panel/content.test.ts:26]: Sorry that I sent you down the TS way, I meant just for the tests, but even those cause many merge problems.

I have reverted the TS transition and squashed our pervious commits into a new PR - as I found it really difficult to review it across the two current PRs that make changes and revert them along the way.

https://github.com/getkirby/kirby/pull/8353

**Comment by distantnative** [comment]
https://github.com/getkirby/kirby/pull/8353
